04-onion-routing.md

```diff
@@ -238,11 +236,47 @@ Field descriptions:
      leaking its position in the route.
 
    * `padding`: This field is for future use and also for ensuring that future non-0-`realm`
-     `per_hop`s won't change the overall `hops_data` size.
+     `hop_data`s won't change the overall `hop_payloads` size.
+
+When forwarding HTLCs, nodes MUST construct the outgoing HTLC as specified
+within `hop_data` above; otherwise, deviation from the specified HTLC
+parameters may lead to extraneous routing failure.
+
+### `tlv_payload` payload format
+
+This is a more flexible format, which avoids the redundant `short_channel_id` field for the final node. 
+
+1. tlv: `tlv_payload`
+2. types:
+    1. type: 0 (`destination_signal`)
```

Comment:
So we have both a signal here and also a final set of zero bytes in the HMAC? I'm not sure we need another signal on this level, given that in the underlying Sphinx format, the zero'd HMAC is used as a tag for the final destination. This would also create two ways that packet processors are required to know of in order to determine if this is the final hop or not. This framing layer sits above the current onion layer IMO, meaning we're crossing layers here in order to detect a terminal point in the route.

Reply:
Or will the HMAC be left in place, and this is just for the higher level TLV parser?

Reply:
I currently kept both conditions in my implementation. If either the next packet's hmac is `0x00` or if the first byte of the TLV stream is `0x00` it means we are the last hop.
Instead of using TLV code inside the onion processor, I translated this condition as `the first byte after the varint length of the payload is 0x00` (since TLV streams are ordered by TLV type values and this one as type `0x00`).
I don't have a strong opinion on what we should do here because it's not necessarily the onion processor layer that need to figure out that we are the destination, it might make more sense that it's the layer that interprets the payload that should figure out that we are the destination.

Reply:
If we leave the HMAC, we should continue to use it to signal termination.

Half-parsing the TLV this way is just nasty, as is the alternative of using type 0xFFFFFFFFFFFFFFFF to signal the HMAC (which, of course would always put it at the end), which loses more bytes across any normal route than skipping the terminal HMAC would!